pkg/gardenlet/controller/shoot/care/add.go

```diff
@@ -0,0 +1,139 @@
+	if r.GardenClient == nil {
+		r.GardenClient = gardenCluster.GetClient()
+	}
+
+	// It's not possible to overwrite the event handler when using the controller builder. Hence, we have to build up
+	// the controller manually.
+	c, err := controller.New(
+		ControllerName,
+		mgr,
+		controller.Options{
+			Reconciler:              r,
+			MaxConcurrentReconciles: pointer.IntDeref(r.Config.Controllers.ShootCare.ConcurrentSyncs, 0),
+			RecoverPanic:            true,
+			// if going into exponential backoff, wait at most the configured sync period
+			RateLimiter: workqueue.NewWithMaxWaitRateLimiter(workqueue.DefaultControllerRateLimiter(), r.Config.Controllers.ShootCare.SyncPeriod.Duration),
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	return c.Watch(
+		source.NewKindWithCache(&gardencorev1beta1.Shoot{}, gardenCluster.GetCache()),
+		r.EventHandler(),
+		predicateutils.SeedNamePredicate(r.SeedName, gutil.GetShootSeedNames),
+		r.ShootPredicate(),
+	)
+}
+
+// RandomDurationWithMetaDuration is an alias for utils.RandomDurationWithMetaDuration.
+var RandomDurationWithMetaDuration = utils.RandomDurationWithMetaDuration
+
+// EventHandler returns a handler for Shoot events.
+func (r *Reconciler) EventHandler() handler.EventHandler {
+	return &handler.Funcs{
+		CreateFunc: func(e event.CreateEvent, q workqueue.RateLimitingInterface) {
+			shoot, ok := e.Object.(*gardencorev1beta1.Shoot)
+			if !ok {
+				return
+			}
+
+			req := reconcile.Request{NamespacedName: types.NamespacedName{
+				Name:      e.Object.GetName(),
+				Namespace: e.Object.GetNamespace(),
+			}}
+
+			if shoot.Generation == shoot.Status.ObservedGeneration {
+				// spread shoot health checks across sync period to avoid checking on all Shoots roughly at the same
+				// time after startup of the gardenlet
+				q.AddAfter(req, RandomDurationWithMetaDuration(r.Config.Controllers.ShootCare.SyncPeriod))
+				return
+			}
+
+			// don't add random duration for enqueueing new Shoots which have never been health checked yet
+			q.Add(req)
+		},
+		UpdateFunc: func(e event.UpdateEvent, q workqueue.RateLimitingInterface) {
+			q.Add(reconcile.Request{NamespacedName: types.NamespacedName{
+				Name:      e.ObjectNew.GetName(),
+				Namespace: e.ObjectNew.GetNamespace(),
+			}})
+		},
+	}
+}
+
+// ShootPredicate is a predicate which returns 'true' for create events, and for update events in case the shoot was
+// successfully reconciled.
+func (r *Reconciler) ShootPredicate() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(event.CreateEvent) bool {
+			return true
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			shoot, ok := e.ObjectNew.(*gardencorev1beta1.Shoot)
+			if !ok {
+				return false
+			}
+
+			oldShoot, ok := e.ObjectOld.(*gardencorev1beta1.Shoot)
+			if !ok {
+				return false
+			}
+
+			// re-evaluate shoot health status right after a reconciliation operation has finished
```

Comment:
`shootReconciliationFinishedSuccessful` function returns true only when the `LastOperation` is Succeeded
```suggestion
			// re-evaluate shoot health status right after a reconciliation operation has succeeded
```